cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java

```diff
@@ -365,30 +378,104 @@ public void apply() throws Exception {
+      return origJobDetail;
+    }
+
+    // Old TriggerKey name has format = namespace:application:type:program
+    String newJobName = getNameWithVersion(splits);
+    return JobBuilder.newJob(DefaultSchedulerService.ScheduledJob.class)
+      .withIdentity(newJobName, origJobDetail.getKey().getGroup())
+      .storeDurably(origJobDetail.isDurable())
+      .build();
+  }
+
+  private boolean addedDefaultAppVersion(TriggerStatusV2 oldTrigger) {
+    TriggerKey oldTriggerKey = oldTrigger.trigger.getKey();
+    JobKey oldTriggerJobKey = oldTrigger.trigger.getJobKey();
+
+    String[] triggerSplits = oldTriggerKey.getName().split(":");
+    String[] oldJobNameSplits = oldTriggerJobKey.getName().split(":");
+
+    // Old TriggerKey name has format = namespace:application:type:program:schedule
+    // Old JobKey name has format = namespace:application:type:program
+    if (triggerSplits.length != 5 && oldJobNameSplits.length != 4) {
+      return false;
+    }
+
+    if (triggerSplits.length == 5) {
+      // This trigger key has the old format. So replace it with new format.
+      // New TriggerKey name has format = namespace:application:version:type:program:schedule
+      String newTriggerName = getNameWithVersion(triggerSplits);
+      oldTrigger.trigger.setKey(new TriggerKey(newTriggerName, oldTriggerKey.getGroup()));
+    }
+
+    if (oldJobNameSplits.length == 4) {
+      // This job key has the old format. So replace it with new format.
+      // New JobKey name has format = namespace:application:version:type:program
+      String newJobName = getNameWithVersion(oldJobNameSplits);
+      oldTrigger.trigger.setJobKey(new JobKey(newJobName, oldTriggerJobKey.getGroup()));
+    }
+    return true;
+  }
+
   private void upgradeJobs() throws Exception {
     Row result = table.get(JOB_KEY);
-
     if (result.isEmpty()) {
       return;
     }
 
     for (byte[] column : result.getColumns().values()) {
-      JobDetail jobDetail = (JobDetail) SerializationUtils.deserialize(column);
-      JobKey oldJobKey = jobDetail.getKey();
-      String oldJobKeyName = oldJobKey.getName();
-      String[] splits = oldJobKeyName.split(":");
-      // Old JobKey name has format = namespace:application:type:program
-      if (splits.length != 4) {
-        LOG.debug("Skip upgrading Job {}. Expected Job key format 'namespace:application:type:program'", oldJobKey);
-        continue;
-      }
+      JobDetail oldJobDetail = (JobDetail) SerializationUtils.deserialize(column);
+      JobDetail jobDetail = addDefaultAppVersionIfNeeded(oldJobDetail);
 
-      // New JobKey name has format = namespace:application:version:type:program
-      String newJobName = getNameWithVersion(splits);
-
-      JobDetail newJobDetail = jobDetail.getJobBuilder().withIdentity(newJobName, oldJobKey.getGroup()).build();
-      persistJob(table, newJobDetail);
-      removeJob(table, oldJobKey);
+      if (!jobDetail.equals(oldJobDetail)) {
+        persistJob(table, jobDetail);
```

Comment:
The same thing happens here too, right? Running an upgrade will override any new updates to a Job or a Trigger.